src/ai/backend/manager/registry.py

```diff
@@ -1273,10 +1272,54 @@ async def _finialize_running() -> None:
                             'stdin_port': created_info['stdin_port'],
                             'stdout_port': created_info['stdout_port'],
                             'service_ports': service_ports,
-                        })
-                        .where(kernels.c.id == created_info['id']))
-                    await conn.execute(query)
-
+                        }
+                        actual_allocs: MutableMapping[str, str] = {}
+                        for alloc_map in created_info['resource_spec']['allocations'].values():
+                            for slot_name, allocations in alloc_map.items():
+                                total_allocs: List[Decimal] = []
+                                for allocation in allocations.values():
+                                    if BinarySize.suffix_map.get(allocation[-1].lower()) is not None:
+                                        total_allocs.append(Decimal(BinarySize.from_str(allocation)))
+                                    else:
+                                        total_allocs.append(Decimal(allocation))
+                                actual_allocs[slot_name] = str(sum(total_allocs))
+
+                        values['occupied_slots'] = actual_allocs
+                        query = (
+                            sa.select([kernels.c.occupied_slots])
+                            .select_from(kernels)
+                            .where(kernels.c.id == created_info['id'])
+                        )
+                        requested_slots = await conn.scalar(query)
+                        query = (
+                            sa.select([agents.c.occupied_slots])
+                            .select_from(agents)
+                            .where(agents.c.id == agent_alloc_ctx.agent_id)
+                        )
+                        agent_occupied_slots = await conn.scalar(query)
+                        alloc_diffs = {
+                            k: Decimal(requested_slots[k]) - Decimal(v)
+                            for k, v in actual_allocs.items()
+                        }
+                        if any([v != 0 for v in alloc_diffs.values()]):
+                            agent_actual_allocated_slots = {
+                                k: str(Decimal(agent_occupied_slots[k]) - v)
+                                for k, v in alloc_diffs.items()
+                            }
+                            query = (
+                                agents.update()
+                                .values({'occupied_slots': agent_actual_allocated_slots})
+                                .where(agents.c.id == agent_alloc_ctx.agent_id)
+                            )
+                            await conn.execute(query)
```

Comment:
Updating kernel-specific resource occupation here looks good because this is a per-kernel "running" finalization handler.  However, for cluster sessions, this would incur multiple concurrent update transactions to the agents table.  How about aggregating and splitting out the agent table update into a single query using a single transaction after updating individual kernels?

Reply:
In that case, how about just calling `AgentRegistry.recalc_resource_usage()` after gathering all agent tasks?